master/internal/command/command.go

```diff
@@ -19,124 +20,116 @@ import (
+		commands map[model.TaskID]*command
+		syslog   *logrus.Entry
 	}
-	return state.Proto()
-}
 
-func createGenericCommandActor(
-	ctx *actor.Context,
-	db *db.PgDB,
-	rm rm.ResourceManager,
-	taskID model.TaskID,
-	taskType model.TaskType,
-	jobID model.JobID,
-	jobType model.JobType,
-	spec *tasks.GenericCommandSpec,
-	contextDirectory []byte,
-) error {
-	spec.TaskType = taskType
-	cmd := &command{
-		db: db,
-		rm: rm,
+	// command is executed in a containerized environment on a Determined cluster.
+	// Locking in: toNTSC, startCmd, onExit, deleteIfInWorkspace.
 
-		GenericCommandSpec: *spec,
+	command struct {
+		mu sync.Mutex
 
-		taskID:   taskID,
-		taskType: taskType,
-		jobType:  jobType,
-		jobID:    jobID,
+		db *db.PgDB
+		rm rm.ResourceManager
 
-		logCtx: logger.Context{
-			"job-id":    jobID,
-			"task-id":   taskID,
-			"task-type": taskType,
-		},
+		tasks.GenericCommandSpec
 
-		contextDirectory: contextDirectory,
+		registeredTime time.Time
+		taskID         model.TaskID
+		taskType       model.TaskType
+		jobType        model.JobType
+		jobID          model.JobID
+		allocationID   model.AllocationID
+		lastState      task.AllocationState
+		exitStatus     *task.AllocationExited
+		restored       bool
+
+		contextDirectory []byte // Don't rely on this being set outsides of PreStart non restore case.
+
+		logCtx logger.Context
+		syslog *logrus.Entry
 	}
 
-	a, _ := ctx.ActorOf(cmd.taskID, cmd)
-	summaryFut := ctx.Ask(a, getSummary{})
-	if err := summaryFut.Error(); err != nil {
-		return errors.Wrap(err, "failed to create generic command")
+	// CreateGeneric is a request to the commandService to create a generic command.
+	CreateGeneric struct {
+		ContextDirectory []byte
+		Spec             *tasks.GenericCommandSpec
 	}
-	// Sync with the actor, but we don't really need the summary. actor.Ping works too,
-	// but this makes sure it can form some sort of useful response (ping doesn't actually
-	// hit the receive block).
-	summaryFut.Get()
-	return nil
-}
+)
 
-func commandFromSnapshot(
-	ctx *actor.Context,
-	db *db.PgDB,
-	rm rm.ResourceManager,
-	snapshot *CommandSnapshot,
-) *command {
-	taskID := snapshot.TaskID
-	taskType := snapshot.Task.TaskType
-	jobID := snapshot.Task.Job.JobID
-	cmd := &command{
-		db:             db,
-		rm:             rm,
-		registeredTime: snapshot.RegisteredTime,
+// newCommandService creates a new commandService.
+func newCommandService() *commandService {
+	return &commandService{
+		commands: make(map[model.TaskID]*command),
+		syslog:   logrus.WithField("component", "command-service"),
+	}
+}
 
-		GenericCommandSpec: snapshot.GenericCommandSpec,
+// InitCommandService starts the global command service singleton.
+func (cs *commandService) InitCommandService(ctx context.Context, db *db.PgDB, rm rm.ResourceManager) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
 
-		taskID:   taskID,
-		taskType: taskType,
-		jobType:  snapshot.Task.Job.JobType,
-		jobID:    jobID,
+	cs.db = db
+	cs.rm = rm
+}
```

Comment:
can we get rid of this and "init" it in `new`? i don't like having a `new` that construct an incorrect/incomplete thing. it is an unintuitive/easily misused API.

Reply:
The issue is that I can't figure out how to initialize/define the command service's db/rm fields _outside_ of calling an Init function in core.go. We need the command service to store these values because each command it creates/tracks needs a db & rm to start its allocation -- and we can't easily find this value when we ask to launch the NTSC.
I'll think some more. I think what you did with allocation service was more simple since you didn't need to pass variables to define the singleton service.

Reply:
can you just make a `NewCommandService(db *db.PgDB, rm rm.ResourceManager) *CommandService` and then `SetCommandService(*CommandService)` to init the singleton? i don't love the pattern i did with allocation, i'll change it.